File: gitdot-web/packages/api/src/endpoint/review/create-review-comments.ts
import { z } from "zod";
import { ReviewResource } from "../../resource";
import type { Endpoint } from "../endpoint";
import { ReviewCommentInput } from "./review-review-diff";

export const CreateReviewCommentsRequest = z.object({
  comments: z.array(ReviewCommentInput),
});
export type CreateReviewCommentsRequest = z.infer<
  typeof CreateReviewCommentsRequest
>;

export const CreateReviewCommentsResponse = ReviewResource;
export type CreateReviewCommentsResponse = z.infer<
  typeof CreateReviewCommentsResponse
>;

export const CreateReviewComments = {
  path: "/repository/{owner}/{repo}/review/{number}/diff/{position}/comments",
  method: "POST",
  request: CreateReviewCommentsRequest,
  response: CreateReviewCommentsResponse,
} as const satisfies Endpoint;
export type CreateReviewComments = typeof CreateReviewComments;

File: gitdot-api/src/endpoint/review/create_review_comments.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{endpoint::Endpoint, resource::review::ReviewResource};

pub struct CreateReviewComments;

impl Endpoint for CreateReviewComments {
    const PATH: &'static str =
        "/repository/{owner}/{repo}/review/{number}/diff/{position}/comments";
    const METHOD: http::Method = http::Method::POST;

    type Request = CreateReviewCommentsRequest;
    type Response = CreateReviewCommentsResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct CreateReviewCommentsRequest {
    pub comments: Vec<ReviewCommentInput>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReviewCommentInput {
    pub revision_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub body: String,
    pub file_path: Option<String>,
    pub line_number_start: Option<i32>,
    pub line_number_end: Option<i32>,
    pub start_character: Option<i32>,
    pub end_character: Option<i32>,
    pub side: Option<String>,
}

pub type CreateReviewCommentsResponse = ReviewResource;
